pkg/securelaunch/helpers.go

```diff
@@ -48,71 +67,232 @@ func WriteToFile(data []byte, dst, defFileName string) (string, error) {
 		Debug("New target=%s", target)
 	}
 
-	Debug("target=%s", target)
+	Debug("WriteToFile: target=%s", target)
 	err = ioutil.WriteFile(target, data, 0644)
 	if err != nil {
 		return "", fmt.Errorf("failed to write date to file =%s, err=%v", target, err)
 	}
-	Debug("WriteToFile exit w success data written to target=%s", target)
+	Debug("WriteToFile: exit w success data written to target=%s", target)
 	return target, nil
 }
 
-/*
- * GetMountedFilePath returns a file path corresponding to a <device_identifier>:<path> user input format.
- * <device_identifier> may be a Linux block device identifier like sda or a FS UUID.
- *
- * NOTE: Caller's responsbility to unmount this..use return var mountPath to unmount in caller.
- */
-func GetMountedFilePath(inputVal string, flags uintptr) (string, string, error) {
-	s := strings.Split(inputVal, ":")
-	if len(s) != 2 {
-		return "", "", fmt.Errorf("%s: Usage: <block device identifier>:<path>", inputVal)
+// persist writes data to targetPath.
+// targetPath is of form sda:/boot/cpuid.txt
+func persist(data []byte, targetPath string, defaultFile string) error {
+
+	filePath, r := GetMountedFilePath(targetPath, 0) // 0 is flag for rw mount option
+	if r != nil {
+		return fmt.Errorf("persist: err: input %s could NOT be located, err=%v", targetPath, r)
+	}
+
+	dst := filePath // /tmp/boot-733276578/cpuid
+
+	target, err := WriteToFile(data, dst, defaultFile)
+	if err != nil {
+		log.Printf("persist: err=%s", err)
+		return err
+	}
+
+	Debug("persist: Target File%s", target)
+	return nil
+}
+
+// AddToPersistQueue enqueues an action item to persistData slice
+// so that it can be deferred to the last step of sluinit.
+func AddToPersistQueue(desc string, data []byte, location string, defFile string) error {
```

Comment:
why is there a queue of these? why isn't this done when it's queued?

Reply:
One of the disadvantages on reviewing code changes in github (this way) is that it doesn't show you commit messages....Let me share that with you here and if you still have questions, I will be happy to answer them...

Here is the individual commit:
https://github.com/u-root/u-root/pull/1559/commits/7eeffca936527d62c2b993645faf1b169f2dd3ac

Reply:
I will try to answer your question again..will make me think also :)
Basically with the queue, the idea is to reduce the number of "remounts". remount is needed if a device that is mounted read-only is to be "remounted" with write permissions(or vice - versa)..
We are deferring the write permissions to make sure first all read operations(like locating policy file, reading files for measurements) are done first and then in the end of sluinit before kexec, we do the write operations...

Note that sluinit mounts a device for write only when it needs to record the measurements on disk for attestation...
Currently, only cpuid collector currently does that but in future we might have other collectors depending on platform needs.

so a good field use case would be if a user wants the securelaunch.policy file to be located on partition XYZ and has bunch of other files that need to be measured on the same parition XYZ and also wants the measurements to be stored on partition XYZ, in that case

our sluinit would (without queue)
1. locate policy file( mount XYZ)
2. measure file 1 (mount XYZ) 
3. measure cpuid and store measurements on XYZ (remount XYZ with write permission)
4. measure file 2 (mount XYZ)
5. write measurements from kernel log buffer (tpmtool parses binary file) to file on XYZ (remount XYZ with write permission)
Total remounts = 2

with queue this order will be
1. locate policy file( mount XYZ)
2. measure file 1 (mount XYZ) 
3 .measure CPUID info and defer writing it 
3. measure file 2 (mount XYZ) 
4. write all measurements (cpuid + others) on XYZ (remount XYZ with write permission)
So only 1 remount... with queue...